src/internal/curry.mjs

```diff
@@ -17,12 +17,49 @@ function unshiftUndefineds (args, by) {
   }
 }
 
+class BaseIterable {
+  constructor (fn, variadic, args, iterableArgs) {
+    this._fn = fn
+    this._args = args
+    this._variadic = variadic
+    this._iterableArgs = iterableArgs
+    this._staticIterator = null
+  }
+
+  _iterate () {
+    return this._variadic ? this._fn(...this._args, this._iterableArgs) : this._fn(...this._args)
+  }
+
+  next () {
+    this._staticIterator = this._staticIterator || this._iterate()
+    return this._staticIterator.next()
+  }
+
+  return (...args) {
+    if (typeof this._staticIterator.return === 'function') this._staticIterator.return(...args)
+  }
+}
+
+class Iterable extends BaseIterable {
+  [Symbol.iterator] () {
+    return this._iterate()
+  }
+}
+
+class AsyncIterable extends BaseIterable {
+  [Symbol.asyncIterator] () {
+    return this._iterate()
+  }
+}
```

Comment:
So instead of returning simple iterators, you create instances of these classes? My problem with this is that this is quite bloated, not to mention that babel often adds extra overhead for class inheritance, this would make JavaScript bundle size larger.

But I also think that you can make it more lightweight by converting class inheritance to object composition.

Reply:
Actually, I think you don't need "composition" part, only "object". What `BaseIterable.constructor` does is merely storing arguments, if you replace these classes with a factory function, you won't be needing those 7 lines of constructor function, you also won't have to use `this`.

Reply:
You are right that the library is most sensitive to bundle size that babel's class transpilation would add here, and that the overhead of classes is not necessary. I have changed this to prototypal inheritance. Does this resolve your concern?